File: modules/Nncase.Modules.CPU/IR/XPU/InstanceNorm.cs
// Copyright (c) Canaan Inc. All rights reserved.
// Licensed under the Apache license. See LICENSE file in the project root for full license information.

namespace Nncase.IR.XPU;

public sealed partial class InstanceNorm : XPUKernelOp
{
    public static readonly ParameterInfo Input = new(typeof(InstanceNorm), 0, "input");

    public static readonly ParameterInfo Scale = new(typeof(InstanceNorm), 1, "scale");

    public static readonly ParameterInfo Bias = new(typeof(InstanceNorm), 2, "bias");

    public static readonly ParameterInfo Output = new(typeof(InstanceNorm), 3, "output");

    public float Epsilon { get; }

    public DistributedType DistType { get; }
}

File: src/ir/ops/instancenorm.cpp
/* Copyright 2019-2021 Canaan Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <nncase/ir/op_utils.h>
#include <nncase/ir/ops/instancenorm.h>
#include <xtensor/xarray.hpp>

using namespace nncase;
using namespace nncase::ir;

instancenorm::instancenorm(datatype_t input_type, shape_t input_shape, float epsilon)
    : epsilon_(epsilon)
{
    add_input("input", input_type, input_shape);
    add_input("scale", input_type, get_instancenorm_const_shape(input_shape));
    add_input("bias", input_type, get_instancenorm_const_shape(input_shape));
    add_output("output", input_type, input_shape);
}

bool instancenorm::properties_equal(node &other) const
{
    auto &r = static_cast<instancenorm &>(other);
    return epsilon() == r.epsilon();
}
